torch/csrc/distributed/c10d/TCPStoreLibUvBackend.cpp

```diff
@@ -16,12 +17,1054 @@
+  }
+
+ public:
+  explicit UvClient(uv_loop_t* loop, LibUVStoreDaemon* store) : store(store) {
+    uv_tcp_init(loop, &client);
+    uv_handle_set_data((uv_handle_t*)&client, this);
+    C10D_DEBUG("Accepted new client: {}\n", (void*)this);
+  }
+
+  void startRead() {
+    int res =
+        uv_read_start((uv_stream_t*)as_stream(), alloc_buffer, read_callback);
+    if (res) {
+      C10D_INFO(
+          "Failed to setup read callback. client:{} code:{} name:{} desc:{}.",
+          (void*)this,
+          res,
+          uv_err_name(res),
+          uv_strerror(res));
+      close();
+    }
+  }
+
+  static UvClient* from_uv(uv_handle_t* handle) {
+    return (UvClient*)uv_handle_get_data(handle);
+  }
+
+  void close() override {
+    if (!uv_is_closing((uv_handle_t*)&client)) {
+      uv_close((uv_handle_t*)&client, on_close);
+    }
+  }
+
+  uv_stream_t* as_stream() override {
+    return (uv_stream_t*)&client;
+  }
+};
+
+void LibUVStoreDaemon::onConnect(int status) {
+  UvClient* client = new UvClient(&loop, this);
+
+  registerClient(client);
+  int res = uv_accept((uv_stream_t*)&server, client->as_stream());
+  if (res == 0) {
+    client->startRead();
+  } else {
+    C10D_INFO(
+        "Failed to accept client. client:{} code:{} name:{} desc:{}.",
+        (void*)client,
+        res,
+        uv_err_name(res),
+        uv_strerror(res));
+    client->close();
+  }
+}
+
+void LibUVStoreDaemon::onExitRequest() {
+  C10D_DEBUG("Store exit requested\n");
+  uv_close((uv_handle_t*)&exit_handle, nullptr);
+  uv_stop(&loop);
+}
+
+uint16_t LibUVStoreDaemon::get_socket_port(uv_tcp_t* handle) {
+  sockaddr_storage addr_s{};
+
+  int addr_len = sizeof(addr_s);
+
+  if (uv_tcp_getsockname(
+          handle, reinterpret_cast<sockaddr*>(&addr_s), &addr_len) != 0) {
+    throw std::runtime_error(
+        "The port number of the socket cannot be retrieved.");
+  }
+
+  if (addr_s.ss_family == AF_INET) {
+    return ntohs(reinterpret_cast<sockaddr_in*>(&addr_s)->sin_port);
+  } else {
+    return ntohs(reinterpret_cast<sockaddr_in6*>(&addr_s)->sin6_port);
+  }
+}
+
+void LibUVStoreDaemon::init(const TCPStoreOptions& opts) {
+  uv_handle_set_data((uv_handle_t*)&server, this);
+
+  if (opts.masterListenFd.has_value()) {
+    int res = uv_tcp_open(&server, *opts.masterListenFd);
+    TORCH_CHECK(
+        res == 0,
+        "Failed to open an existing socket. code:{} name:{} message:{}",
+        res,
+        uv_err_name(res),
+        uv_strerror(res));
+
+    port_ = get_socket_port(&server);
+    TORCH_CHECK(
+        port_ == opts.port,
+        "listen fd {} is bound to port {}, expected to be bound to port {}",
+        *opts.masterListenFd,
+        port_,
+        opts.port);
+
+    return;
+  }
+
+  if (tryListen(true)) {
+    return;
+  }
+  uv_close((uv_handle_t*)&server, nullptr);
+  uv_run(&loop, UV_RUN_ONCE);
+
+  if (tryListen(false)) {
+    return;
+  }
+  uv_close((uv_handle_t*)&server, nullptr);
+  uv_run(&loop, UV_RUN_ONCE);
```

Comment:
nit: can this be updated to be?
```C++
  if (tryListen(true) || tryListen(false)) {
    return;
  }
  uv_close((uv_handle_t*)&server, nullptr);
  uv_run(&loop, UV_RUN_ONCE);
```

I'm not familiar with the reason for calling uv_close and uv_run twice

Reply:
This part of the code is really really fuggly.
We need to call close otherwise we leak the socket as tryListen creates a new one.
We need to call `uv_run` as operations are asynchronous and we can't reuse the handle memory until the uv loop processes and closes it.